File: examples/demo_eldc_lib.java
/**
 * demo_eldc_lib.java — ELDC shared library demo for Java (JNA).
 *
 * Dependency: jna-5.x.jar  (https://github.com/java-native-access/jna)
 * Compile & run:
 *   javac -cp jna.jar demo_eldc_lib.java
 *   java  -cp .:jna.jar -Djna.library.path=. demo_eldc_lib
 */

import com.sun.jna.*;
import java.util.Arrays;

public class demo_eldc_lib {

    static final int MAX_SCORES = 20;

    @Structure.FieldOrder({"language", "score"})
    public static class EldcScoreItem extends Structure {
        public Pointer language;   // const char* — read with .getString(0)
        public float   score;
    }

    @Structure.FieldOrder({"language", "reliable", "n_scores", "scores"})
    public static class EldcDetectResult extends Structure {
        public Pointer        language;
        public int            reliable;
        public int            n_scores;
        public EldcScoreItem[] scores = (EldcScoreItem[]) new EldcScoreItem().toArray(MAX_SCORES);
    }

    public interface EldcLib extends Library {
        EldcLib LIB = Native.load("eldc", EldcLib.class);

        // ── Global API ───────────────────────────────────────────────────────
        void    eldc_init();
        void    eldc_close();
        String  eldc_detect(String text);
        void    eldc_detect_details(String text, EldcDetectResult result);
        String  eldc_set_languages(String codes);
        void    eldc_set_scheme(String scheme);
        void    eldc_set_scores(int n);

        // ── Instance (cfg) API — EldcConfig is opaque, use Pointer ──────────
        Pointer eldc_config_create();
        void    eldc_config_free(Pointer cfg);
        String  eldc_detect_cfg(Pointer cfg, String text);
        void    eldc_detect_details_cfg(Pointer cfg, String text, EldcDetectResult result);
        String  eldc_set_languages_cfg(Pointer cfg, String codes);
        void    eldc_set_scheme_cfg(Pointer cfg, String scheme);
        void    eldc_set_scores_cfg(Pointer cfg, int n);
    }

    static String lang(Pointer p) { return p != null ? p.getString(0) : "und"; }

    public static void main(String[] args) {
        var lib = EldcLib.LIB;

        // ── 1. init ──────────────────────────────────────────────────────────
        lib.eldc_init();
        System.out.println("=== eldc Java demo ===\n");
	     // ── Optional: Isolated configuration instance ────────────────────────
		  var i_config = lib.eldc_config_create();

        // ── 2. detect ────────────────────────────────────────────────────────
        System.out.println("-- detect --");
        System.out.println(lib.eldc_detect("Bonjour le monde"));  // fr
        System.out.println(lib.eldc_detect("12345 !@#"));         // und
		  // Same behavior for instance protected configuration detect
		  System.out.println(lib.eldc_detect_cfg(i_config , "Bonjour"));  // fr

        // ── 3. detect_details ────────────────────────────────────────────────
        // eldc_detect_details fills the struct and returns void; read from r.
        System.out.println("\n-- detect_details --");
        var r = new EldcDetectResult();
        lib.eldc_detect_details("Bonjour le monde", r);
        System.out.printf("language : %s  reliable: %b%n", lang(r.language), r.reliable == 1);
        for (int i = 0; i < r.n_scores; i++)
            System.out.printf("  %s: %.4f%n", lang(r.scores[i].language), r.scores[i].score);

		  lib.eldc_detect_details_cfg(i_config, "Bonjour", r);  // Same behavior

        // ── 4. set_scores ─────────────────────────────────────────────────────
        System.out.println("\n-- set_scores(2) --");
        lib.eldc_set_scores(2);  // Default 3, Max 20, Min 1. Global setter
        var r2 = new EldcDetectResult();
        lib.eldc_detect_details("Was ist das?", r2);
        System.out.printf("language: %s  n_scores: %d%n", lang(r2.language), r2.n_scores);  // de, 2
        lib.eldc_set_scores(3);  // reset

		  lib.eldc_set_scores_cfg(i_config, 2);  // instance setter

        // ── 5. set_languages ──────────────────────────────────────────────────
        System.out.println("\n-- set_languages --");
        System.out.println(lib.eldc_set_languages("en,fr,es"));    // en,fr,es  Global setter
        System.out.println(lib.eldc_set_languages("eng,fra,xyz")); // en,fr (xyz → stderr)
        System.out.println(lib.eldc_detect("Bonjour le monde"));   // fr
        lib.eldc_set_languages("");                                // reset
        System.out.println(lib.eldc_detect("Hola doctor"));        // es

		  lib.eldc_set_languages_cfg(i_config, "en,fr,es");          // instance setter

        // ── 6. set_scheme ─────────────────────────────────────────────────────
        System.out.println("\n-- set_scheme --");
        lib.eldc_set_scheme("iso639-2t");                // Global setter
        System.out.println(lib.eldc_detect("Bonjour"));  // fra
        lib.eldc_set_scheme("iso639-1");
        System.out.println(lib.eldc_detect("Bonjour"));  // fr

		  lib.eldc_set_scheme_cfg(i_config, "iso639-2t");  // instance setteR

        // ── 7. cleanup ────────────────────────────────────────────────────────
        lib.eldc_config_free(i_config);
        // Global unload
        lib.eldc_close();
    }
}

File: examples/demo_eldc_lib.ts
/**
 * demo_eldc_lib.ts — ELDC shared library demo for Node.js / TypeScript (koffi).
 *
 * npm install koffi
 * Run: npx ts-node demo_eldc_lib.ts
 *  or: node --loader ts-node/esm demo_eldc_lib.ts
 *
 * Also works with plain JS: rename to .js and remove the type annotations.
 */

import koffi from 'koffi';
import { join } from 'path';

const libName = process.platform === 'win32'  ? 'eldc.dll'
              : process.platform === 'darwin' ? join(__dirname, 'libeldc.dylib')
              :                                 join(__dirname, 'libeldc.so');

const lib = koffi.load(libName);

// ── Struct definitions ────────────────────────────────────────────────────────
const MAX_SCORES = 20;

const EldcScoreItem = koffi.struct('EldcScoreItem', {
    language: 'const char *',
    score:    'float',
});

const EldcDetectResult = koffi.struct('EldcDetectResult', {
    language: 'const char *',
    reliable: 'int',
    n_scores: 'int',
    scores:   koffi.array(EldcScoreItem, MAX_SCORES),
});

// ── Global API bindings ───────────────────────────────────────────────────────
const eldc_init           = lib.func('void eldc_init()');
const eldc_close          = lib.func('void eldc_close()');
const eldc_detect         = lib.func('const char *eldc_detect(const char *text)');
const eldc_detect_details = lib.func('void eldc_detect_details(const char *text, _Out_ EldcDetectResult *result)');
const eldc_set_languages  = lib.func('const char *eldc_set_languages(const char *codes)');
const eldc_set_scheme     = lib.func('void eldc_set_scheme(const char *scheme)');
const eldc_set_scores     = lib.func('void eldc_set_scores(int n)');

// ── Instance (cfg) API bindings — EldcConfig is opaque, use void * ───────────
const eldc_config_create      = lib.func('void *eldc_config_create()');
const eldc_config_free        = lib.func('void eldc_config_free(void *cfg)');
const eldc_detect_cfg         = lib.func('const char *eldc_detect_cfg(void *cfg, const char *text)');
const eldc_detect_details_cfg = lib.func('void eldc_detect_details_cfg(void *cfg, const char *text, _Out_ EldcDetectResult *result)');
const eldc_set_languages_cfg  = lib.func('const char *eldc_set_languages_cfg(void *cfg, const char *codes)');
const eldc_set_scheme_cfg     = lib.func('void eldc_set_scheme_cfg(void *cfg, const char *scheme)');
const eldc_set_scores_cfg     = lib.func('void eldc_set_scores_cfg(void *cfg, int n)');

// ── 1. init ──────────────────────────────────────────────────────────────────
eldc_init();
// ── Optional: Isolated configuration instance ────────────────────────────────
const i_config = eldc_config_create();

console.log('=== eldc TypeScript/Node.js demo ===\n');

// ── 2. detect ────────────────────────────────────────────────────────────────
console.log('-- detect --');
console.log(eldc_detect('Bonjour le monde'));   // fr
console.log(eldc_detect('12345 !@#'));          // und

console.log(eldc_detect_cfg(i_config, 'Bonjour'));  // fr, Same behavior

// ── 3. detect_details ────────────────────────────────────────────────────────
// eldc_detect_details fills the struct and returns void; read result from r.
console.log('\n-- detect_details --');
const r: any = {};
eldc_detect_details('Bonjour le monde', r);
console.log(`language : ${r.language}  reliable: ${r.reliable === 1}`);
for (let i = 0; i < r.n_scores; i++)
    console.log(`  ${r.scores[i].language}: ${r.scores[i].score.toFixed(4)}`);
 
eldc_detect_details_cfg(i_config, 'Bonjour', r);  // Same behavior

// ── 4. set_scores ─────────────────────────────────────────────────────────────
console.log('\n-- set_scores(2) --');
eldc_set_scores(2);   // Default 3, Max 20, Min 1. Global setter
const r2: any = {};
eldc_detect_details('Was ist das?', r2);
console.log(`language: ${r2.language}  n_scores: ${r2.n_scores}`);  // de, 2
eldc_set_scores(3);   // reset

eldc_set_scores_cfg(i_config, 2);  // instance setter

// ── 5. set_languages ──────────────────────────────────────────────────────────
console.log('\n-- set_languages --');
console.log(eldc_set_languages('en,fr,es'));     // en,fr,es  Global setter
console.log(eldc_set_languages('eng,fra,xyz'));  // en,fr  (xyz → stderr)
console.log(eldc_detect('Bonjour le monde'));    // fr
eldc_set_languages('');                          // reset
console.log(eldc_detect('Hola doctor'));         // es

eldc_set_languages_cfg(i_config, 'en,fr,es');    // instance setter

// ── 6. set_scheme ─────────────────────────────────────────────────────────────
console.log('\n-- set_scheme --');
eldc_set_scheme('iso639-2t');                    // Global setter
console.log(eldc_detect('Bonjour'));             // fra
eldc_set_scheme('iso639-1');
console.log(eldc_detect('Bonjour'));             // fr

eldc_set_scheme_cfg(i_config, 'iso639-2t');      // instance setter

// ── 7. Cleanup ─────────────────────────────────────────────────────────
eldc_config_free(i_config);
// Global unload
eldc_close();
